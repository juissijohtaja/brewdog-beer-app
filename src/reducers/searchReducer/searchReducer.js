const searchReducer = (state = '', action) => {
  switch (action.type) {
  case 'SET_KEYWORD':
    return action.keyword
  case 'RESET_KEYWORD':
    return ''
  default:
    return state
  }
}

export const keywordChange = keyword => ({ type: 'SET_KEYWORD', keyword })
export const keywordReset = () => ({ type: 'RESET_KEYWORD' })

export default searchReducer