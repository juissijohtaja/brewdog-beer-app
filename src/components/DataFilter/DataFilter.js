import React, { useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import InputAdornment from '@material-ui/core/InputAdornment'
import IconButton from '@material-ui/core/IconButton'
import CloseIcon from '@material-ui/icons/Close'
import Box from '@material-ui/core/Box'
import SearchIcon from '@material-ui/icons/Search'
import { makeStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'
import Chip from '@material-ui/core/Chip'
import { keywordChange, keywordReset } from '../../reducers/searchReducer/searchReducer'

const useStyles = makeStyles(() => ({
  root: {
    backgroundColor: 'white',
    borderRadius: 4,
  },
  chipSpacing: {
    marginRight: 6
  }
}))

const DataFilter = () => {
  const dispatch = useDispatch()
  const keyword = useSelector(state => state.keyword)
  const classes = useStyles()
  const [tag, setTag] = useState('')

  const handleFilterChange = (event) => {
    dispatch(keywordChange(event.target.value))
  }
  const handleFilterClear = () => {
    dispatch(keywordReset())
  }

  const handleClickTag = (value) => {
    console.info('You clicked the Chip.')
    setTag(value)
    dispatch(keywordChange(value))
  }
  console.log('tag', tag)

  return (
    <Box marginTop={1} marginBottom={2} data-testid='data-filter'>
      <TextField
        className={classes.root}
        value={keyword}
        variant='outlined'
        onChange={handleFilterChange}
        id='searchField'
        aria-label='search'
        placeholder='Search by food pairing...'
        fullWidth
        margin='normal'
        InputProps={{
          endAdornment: (
            <InputAdornment position='end'>
              <IconButton
                aria-label='clear filter'
                onClick={handleFilterClear}
              >
                {keyword ? <CloseIcon /> : null }
              </IconButton>
            </InputAdornment>
          ),
          startAdornment: (
            <InputAdornment position='start'>
              <SearchIcon />
            </InputAdornment>
          ),
        }}
      />
      <Box>
        <Chip
          label='chicken'
          color='primary'
          onClick={() => handleClickTag('chicken')}
          className={classes.chipSpacing}
        />
        <Chip
          label='crab'
          color='primary'
          onClick={() => handleClickTag('crab')}
          className={classes.chipSpacing}
        />
        <Chip
          label='burger'
          color='primary'
          onClick={() => handleClickTag('burger')}
          className={classes.chipSpacing}
        />
        <Chip
          label='spicy'
          color='primary'
          onClick={() => handleClickTag('spicy')}
          className={classes.chipSpacing}
        />
        <Chip
          label='salad'
          color='primary'
          onClick={() => handleClickTag('salad')}
          className={classes.chipSpacing}
        />
      </Box>
    </Box>
  )
}

export default DataFilter