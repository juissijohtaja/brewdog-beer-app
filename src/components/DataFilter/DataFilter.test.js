import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import initialStoreState from '../../test-utils'
import DataFilter from './DataFilter'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { keywordChange, keywordReset } from '../../reducers/searchReducer/searchReducer'

describe('DataFilter Component', () => {
  let store, component
  const mockStore = configureStore([])

  beforeEach(() => {
    store = mockStore({
      initialStoreState
    })

    store.dispatch = jest.fn()

    component = renderer.create(
      <Provider store={store}>
        <DataFilter />
      </Provider>
    )
  })

  test('search field should render empty Search field, SearchIcon and placeholder text', () => {
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('clear search field should dispatch action', () => {
    renderer.act(() => {
      component.root.findByType('button').props.onClick()
    })

    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith(
      keywordReset()
    )
  })

  test('input on search field should dispatch action', () => {
    renderer.act(() => {
      component.root.findByType('input')
        .props.onChange({ target: { value: 'c' } })
    })

    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith(
      keywordChange('c')
    )
  })
})