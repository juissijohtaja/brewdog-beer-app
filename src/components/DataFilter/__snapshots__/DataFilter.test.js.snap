// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataFilter Component search field should render empty Search field, SearchIcon and placeholder text 1`] = `
<div
  className="MuiBox-root MuiBox-root-2"
  data-testid="data-filter"
>
  <div
    aria-label="search"
    className="MuiFormControl-root MuiTextField-root makeStyles-root-1 MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
      onClick={[Function]}
    >
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionStart"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </div>
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
        disabled={false}
        id="searchField"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search..."
        required={false}
        type="text"
      />
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
      >
        <button
          aria-label="clear filter"
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          />
        </button>
      </div>
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        style={
          Object {
            "paddingLeft": 8,
          }
        }
      >
        <legend
          className="PrivateNotchedOutline-legend-4"
          style={
            Object {
              "width": 0.01,
            }
          }
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
